crates/ibc/src/core/ics02_client/handler/misbehaviour.rs

```diff
@@ -0,0 +1,391 @@
+use crate::core::ics02_client::events::ClientMisbehaviour;
+use crate::core::ics02_client::handler::ClientResult;
+use crate::core::ics02_client::msgs::misbehaviour::MsgSubmitMisbehaviour;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::events::IbcEvent;
+use crate::handler::{HandlerOutput, HandlerResult};
+use crate::prelude::*;
+
+/// The result following the successful processing of a `MsgSubmitMisbehaviour` message.
+#[derive(Clone, Debug, PartialEq)]
+pub struct MisbehaviourResult {
+    pub client_id: ClientId,
+    pub client_state: Box<dyn ClientState>,
+}
+
+pub fn process(
+    ctx: &dyn ClientReader,
+    msg: MsgSubmitMisbehaviour,
+) -> HandlerResult<ClientResult, ClientError> {
+    let mut output = HandlerOutput::builder();
+
+    let MsgSubmitMisbehaviour {
+        client_id,
+        misbehaviour,
+        signer: _,
+    } = msg;
+
+    // Read client state from the host chain store.
+    let client_state = ctx.client_state(&client_id)?;
+
+    if client_state.is_frozen() {
+        return Err(ClientError::ClientFrozen { client_id });
+    }
+
+    let client_state = client_state
+        .check_misbehaviour_and_update_state(ctx, client_id.clone(), misbehaviour)
+        .map_err(|e| ClientError::MisbehaviourHandlingFailure {
+            reason: e.to_string(),
+        })?;
+
+    output.emit(IbcEvent::ClientMisbehaviour(ClientMisbehaviour::new(
+        client_id.clone(),
+        client_state.client_type(),
+    )));
+
+    let result = ClientResult::Misbehaviour(MisbehaviourResult {
+        client_id,
+        client_state,
+    });
+
+    Ok(output.with_result(result))
+}
+
+#[cfg(test)]
+mod tests {
+    use core::str::FromStr;
+    use test_log::test;
+
+    use crate::clients::ics07_tendermint::client_type as tm_client_type;
+    use crate::clients::ics07_tendermint::header::Header as TmHeader;
+    use crate::clients::ics07_tendermint::misbehaviour::Misbehaviour as TmMisbehaviour;
+    use crate::core::ics02_client::client_state::ClientState;
+    use crate::core::ics02_client::error::ClientError;
+    use crate::core::ics02_client::handler::dispatch;
+    use crate::core::ics02_client::handler::ClientResult;
+    use crate::core::ics02_client::msgs::misbehaviour::MsgSubmitMisbehaviour;
+    use crate::core::ics02_client::msgs::ClientMsg;
+    use crate::core::ics24_host::identifier::{ChainId, ClientId};
+    use crate::events::IbcEvent;
+    use crate::handler::HandlerOutput;
+    use crate::mock::client_state::client_type as mock_client_type;
+    use crate::mock::client_state::MockClientState;
+    use crate::mock::context::MockContext;
+    use crate::mock::header::MockHeader;
+    use crate::mock::host::{HostBlock, HostType};
+    use crate::mock::misbehaviour::Misbehaviour;
+    use crate::test_utils::get_dummy_account_id;
+    use crate::timestamp::Timestamp;
+    use crate::Height;
+    use crate::{downcast, prelude::*};
+
+    #[test]
+    fn test_misbehaviour_client_ok() {
```

Comment:
great tests! could you please add descriptions if the two headers are different and if yes how they differ? It is not clear especially for the synthetic headers.
Also explain why sometimes we expect empty Log and other times misbehaviour event.

Reply:
Sorry for not doing this before. Did some major cleanup and added comments -> https://github.com/cosmos/ibc-rs/pull/215/files/ebdd644ffa0952e6d9a80e4096008c992ea5e4bd..0780ad60b0acbecad2b08ef46dc787861cbce806